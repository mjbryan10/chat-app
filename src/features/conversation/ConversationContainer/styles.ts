import styled, { ThemeProps, Theme } from 'styled-components';
import { breakpoint } from 'shared/styles/Variables';

interface ButtonProps extends ThemeProps<Theme> {
   isCreateMode?: boolean;
}

export const Container = styled.div`
   display: flex;
   flex-flow: column nowrap;
   align-items: center;
   width: 100%;
   min-height: 60vh;
   height: 100%;
   overflow-y: auto;
   @media ${breakpoint.tablet} {
      max-width: 400px;
      min-width: 300px;
   }
`;

export const Header = styled.h2<ThemeProps<Theme>>`
   text-transform: uppercase;
   text-align: center;
   color: ${({ theme }) => theme.color.text ?? 'inherit'};
   margin: 1rem auto;
`;

export const Button = styled.button<ButtonProps>`
   margin-top: auto;
   padding: 1rem;
   border: ${({ theme }) => theme.border};
   display: flex;
   justify-content: center;
   align-items: center;
   background: ${({ theme, isCreateMode }) =>
      isCreateMode ? theme.background.buttonPrimary : theme.background.conversation};
   color: ${({ theme, isCreateMode }) =>
      isCreateMode ? theme.color.white : theme.color.text};
   &:active {
      background: ${({ theme, isCreateMode }) =>
         isCreateMode
            ? theme.background.buttonPrimary
            : theme.background.conversationSelected};
   }
   &:hover {
      cursor: pointer;
   }
`;
