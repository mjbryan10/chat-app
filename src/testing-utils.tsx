import React from 'react';
import { render } from '@testing-library/react'
import { ThemeProvider } from 'styled-components'
import { dark } from './shared/theme';

const AllTheProviders = ({ children }: any) => {
  return (
    <ThemeProvider theme={dark}>
        {children}
    </ThemeProvider>
  )
  
}

const customRender = (ui?: any, options?: any) =>
  render(ui, { wrapper: AllTheProviders, ...options })

// re-export everything
export * from '@testing-library/react'

// override render method
export { customRender as render }