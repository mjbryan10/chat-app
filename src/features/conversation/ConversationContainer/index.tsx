import React, { useEffect, SyntheticEvent, useState } from 'react';
import * as S from './styles';
import { useSelector, useDispatch } from 'react-redux';
import {
   selectConversations,
   fetchConversations,
   clearConversations,
   selectConversationStatus,
} from '../conversationSlice';
import { selectLogin, logoutUser } from 'features/login/loginSlice';
import { AiOutlinePlusSquare, AiOutlineCheckSquare } from 'react-icons/ai';
import ConversationSelector from '../ConversationSelector';
import ConversationUserSelector from '../ConversationUserSelector';
import {
   selectSelectedUsers,
   clearSelectedUsers,
   clearSelectableUsers,
} from 'features/users/usersSlice';
import ConversationApi from 'shared/Api/ConversationApi';
import InputTextFieldWithLabel from 'components/InputTextFieldWithLabel';
import {
   setNavigationState,
   selectNavigationState,
} from 'features/navigation/navigationSlice';

const ConversationContainer = () => {
   //REDUX:
   const conversations = useSelector(selectConversations);
   const { id: currentUserId } = useSelector(selectLogin);
   const status = useSelector(selectConversationStatus);
   const selectedUsers = useSelector(selectSelectedUsers);
   const navigationState = useSelector(selectNavigationState);
   const dispatch = useDispatch();

   //STATE:
   /**
    * State for managing if container should render conversation
    * selection or creation components.
    */
   const [isCreateMode, setIsCreateMode] = useState(false);
   const [error, setError] = useState('');
   const [conversationName, setConversationName] = useState('');
   /**
    * Handles the fetching and clearing of conversations
    */

   //CYCLE EFFECTS:

   /**
    * Fetches new conversations.
    */
   useEffect(() => {
      if (currentUserId && !isCreateMode) {
         dispatch(fetchConversations(currentUserId));
      }
      return () => {
         dispatch(clearConversations);
      };
   }, [currentUserId, dispatch, isCreateMode]);
   
   /**
    * Subscribes to the nagivationState and reacts when is set to index.
    * 
    * This allows user to use navigation tools to escape creation process.
    */
   useEffect(() => {
      if (navigationState === 'index' && isCreateMode) {
         dispatch(clearSelectedUsers());
         dispatch(clearSelectableUsers());
         setError('');
         setIsCreateMode(false);
      }
   }, [dispatch, isCreateMode, navigationState]);

   //METHODS:

   /**
    * toggles the create mode state. If the new mode is to be the create mode
    * it informs redux of the new navigation state.
    */
   const toggleCreateMode = () => {
      const newMode = !isCreateMode;
      setError('');
      setIsCreateMode(newMode);
      if (newMode) {
         dispatch(setNavigationState('conversation'));
         dispatch(clearConversations());
      }
   };
   /**
    * Validates the creation process, updating the error state where
    * appropiate
    */
   const validateCreation = (): boolean => {
      if (!currentUserId) {
         dispatch(logoutUser());
         return false;
      }
      if (!selectedUsers.length) {
         setError('Select who you wish to speak with first');
         return false;
      }
      if (selectedUsers.length >= 2 && !conversationName.length) {
         setError('Group conversations need a name');
         return false;
      }
      return true;
   };
   /**
    *
    */
   const createConversation = () => {
      if (!currentUserId || !validateCreation()) return;
      //Shorcuts and does not process if cannot validate.
      const conversationApi = new ConversationApi();
      const userIds: number[] = selectedUsers.map((user) => user.id);
      userIds.push(currentUserId);

      if (selectedUsers.length === 1) {
         conversationApi
            .createNewPersonalConversation(userIds)
            .then((res) => {
               resolveCreation();
            })
            .catch((error) => {
               console.error(error);
               setError('Oops, something went wrong. Please try again later.');
            });
      } else if (selectedUsers.length >= 2) {
         conversationApi
            .createNewGroupConversation(userIds, conversationName)
            .then((res) => {
               resolveCreation();
            })
            .catch((error) => {
               console.error(error);
               setError('Oops, something went wrong. Please try again later.');
            });
      }
   };
   const resolveCreation = () => {
      dispatch(clearSelectedUsers());
      dispatch(clearSelectableUsers());
      setConversationName('');
      toggleCreateMode();
      if (currentUserId) {
         dispatch(fetchConversations(currentUserId));
      }
   };
   /**
    * Handles the button clicking depending on if it is in create mode
    * @param event Button click event
    */
   const onClick = (event: SyntheticEvent<HTMLButtonElement>) => {
      event.preventDefault();
      if (!isCreateMode) {
         toggleCreateMode();
         return;
      }
      createConversation();
   };

   const handleConversationNameChange = (value: string) => {
      setError('');
      setConversationName(value);
   };
   return (
      <S.Container>
         <S.Header>Conversations</S.Header>
         <S.Display>
            {isCreateMode ? (
               <ConversationUserSelector />
            ) : (
               <ConversationSelector status={status} conversations={conversations} data-testid="conversation-selector" />
            )}
         </S.Display>
         <S.ControlsContainer>
            {selectedUsers.length >= 2 ? (
               <InputTextFieldWithLabel
                  labelValue="Group Name"
                  handleChange={handleConversationNameChange}
                  data-testid="conversation-name-input"
               />
            ) : null}
            {error.length ? <S.ErrorText>{error}</S.ErrorText> : null}
            <S.Button onClick={onClick} isCreateMode={isCreateMode}>
               {isCreateMode ? (
                  <AiOutlineCheckSquare size="3rem" />
               ) : (
                  <AiOutlinePlusSquare size="3rem" />
               )}
               Create new conversation
            </S.Button>
         </S.ControlsContainer>
      </S.Container>
   );
};

export default ConversationContainer;
