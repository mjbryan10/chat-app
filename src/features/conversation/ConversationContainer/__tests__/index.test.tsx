import React from 'react';
import ConversationContainer from '..';
import { render } from 'testing-utils';

describe('ConversationContainer Feature Component', () => {
  test('should have a create conversation button', () => {
    const utils = render(<ConversationContainer />);
    const {getByText} = utils;


    expect(getByText(/create new conversation/i)).toBeVisible();
  })
  
})
