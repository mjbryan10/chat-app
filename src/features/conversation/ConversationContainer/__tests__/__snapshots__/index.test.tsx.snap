// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ConversationContainer Feature Component should render correctly 1`] = `
<DocumentFragment>
  <div
    class="sc-AxjAm IORMA"
  >
    <h2
      class="sc-AxirZ cyzKdl"
    >
      Conversations
    </h2>
    <div
      class="sc-AxgMl LrMgL"
    />
    <div
      class="sc-AxhCb gzPord"
    >
      <button
        class="sc-AxiKw dTKzca"
      >
        <svg
          fill="currentColor"
          height="3rem"
          stroke="currentColor"
          stroke-width="0"
          viewBox="0 0 1024 1024"
          width="3rem"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M328 544h152v152c0 4.4 3.6 8 8 8h48c4.4 0 8-3.6 8-8V544h152c4.4 0 8-3.6 8-8v-48c0-4.4-3.6-8-8-8H544V328c0-4.4-3.6-8-8-8h-48c-4.4 0-8 3.6-8 8v152H328c-4.4 0-8 3.6-8 8v48c0 4.4 3.6 8 8 8z"
          />
          <path
            d="M880 112H144c-17.7 0-32 14.3-32 32v736c0 17.7 14.3 32 32 32h736c17.7 0 32-14.3 32-32V144c0-17.7-14.3-32-32-32zm-40 728H184V184h656v656z"
          />
        </svg>
        Create new conversation
      </button>
    </div>
  </div>
</DocumentFragment>
`;
